src/runtime_src/core/edge/skd/xrt_skd.cpp

```diff
@@ -35,313 +40,378 @@ namespace xrt {
-    kernel = dlsym(sk_handle, sk_name);
-    errstr = dlerror();
-    if(errstr != NULL) {
-      syslog(LOG_ERR, "Dynamic Link error: %s\n", errstr);
-      return -1;
-    }
-    if (!kernel) {
-      syslog(LOG_ERR, "Cannot find kernel %s\n", sk_name);
-      return -1;
+    m_kernel = dlsym(m_sk_handle, m_sk_name.c_str());
+    if (m_kernel == nullptr) {
+      const std::string errstr = dlerror();
+      const auto errMsg = boost::format("Dynamic Link error: %s - Cannot find kernel %s") % errstr % m_sk_name;
+      xrt_core::message::send(severity_level::error, "SKD", errMsg.str());
+      return -ELIBACC;
     }
 
     // Soft kernel command bohandle init
-    ret = createSoftKernel(&cmd_boh);
+    ret = create_softkernel(&m_cmd_boh);
     if (ret) {
-      syslog(LOG_ERR, "Cannot create soft kernel.");
-      return -1;
+      const auto errMsg = "Cannot create soft kernel.";
+      xrt_core::message::send(severity_level::error, "SKD", errMsg);
+      return ret;
     }
 
-    syslog(LOG_INFO, "%s_%d start running, cmd_boh = %d\n", sk_name, cu_idx, cmd_boh);
+    const auto msg4 = boost::format("%s%s start running, cmd_boh = %s") % m_sk_name % std::to_string(m_cu_idx) % std::to_string(m_cmd_boh);
+    xrt_core::message::send(severity_level::info, "SKD", msg4.str());
 
-    args_from_host = (unsigned *)xclMapBO(devHdl, cmd_boh, true);;
-    if (args_from_host == MAP_FAILED) {
-      syslog(LOG_ERR, "Failed to map soft kernel args for %s_%d", sk_name, cu_idx);
-      dlclose(sk_handle);
-      return -1;
+    m_args_from_host = reinterpret_cast<unsigned *>(xclMapBO(m_devhdl, m_cmd_boh, true));
+    if (m_args_from_host == MAP_FAILED) {
+      const auto errMsg = boost::format("Failed to map soft kernel args for %s%s") % m_sk_name % std::to_string(m_cu_idx);
+      xrt_core::message::send(severity_level::error, "SKD", errMsg.str());
+      dlclose(m_sk_handle);
+      return -EINVAL;
     }
 
     // Prep FFI type for all kernel arguments
-    for(int i=0;i<num_args;i++) {
-      ffi_args[i] = convert_to_ffitype(args[i]);
+    for(const auto &i : m_kernel_args) {
+      m_ffi_args.emplace_back(convert_to_ffitype(i));
     }
 
-    if(ffi_prep_cif(&cif,FFI_DEFAULT_ABI, num_args, &ffi_type_uint32,ffi_args) != FFI_OK) {
-      syslog(LOG_ERR, "Cannot prep FFI arguments!");
-      return -1;
+    // Expect PS kernels to return POSIX return code 
+    if(ffi_prep_cif(&m_cif,FFI_DEFAULT_ABI, m_kernel_args.size(), &ffi_type_uint32, m_ffi_args.data()) != FFI_OK) {
+      const auto errMsg = "Cannot prep FFI arguments!";
+      xrt_core::message::send(severity_level::error, "SKD", errMsg);
+      return -EINVAL;
     }
 
-    syslog(LOG_INFO,"Finish soft kernel %s init\n",sk_name);
+
+    const auto msg5 = boost::format("Finish soft kernel %s init") % m_sk_name;
+    xrt_core::message::send(severity_level::debug, "SKD", msg5.str());
     return 0;
   }
 
   XCL_DRIVER_DLLESPEC
   void
   skd::run() {
-    int32_t kernel_return = 0;
```

Comment:
Should this return error code or does this throw exception in case of errors?

Reply:
This is the kernel return value back to the host.
Right now only xma is looking at this return value.